synapse/storage/databases/main/push_rule.py

```diff
@@ -396,10 +396,6 @@ def _add_push_rule_relative_txn(
         before: str,
         after: str,
     ) -> None:
-        # Lock the table since otherwise we'll have annoying races between the
-        # SELECT here and the UPSERT below.
-        self.database_engine.lock_table(txn, "push_rules")
```

Comment:
I wanted to describe the race to ensure we have a shared understanding. Given two rules `rule_A` and `rule_B` with priorities 0 and 1, respectively. If you make two requests:

1. Add a new push rule (`rule_X`) after `rule_A`, this should make `rule_B` a priority of 2; `rule_X` priority 1; and leave `rule_A` at 0.
2. Add a new push rule (`rule_Y`) after `rule_A`, this should make `rule_B` a priority of 3; `rule_X` priority 2; `rule_Y` priority 1; and leave `rule_A` at 0.

This is in a transaction (I assume running at [`READ COMMITTED`](https://www.postgresql.org/docs/current/transaction-iso.html#XACT-READ-COMMITTED)), so what happens if these race?

Reply:
I believe the winner will be applied and the second transaction will be replayed using the new updated data, per (added some breaks to make it easier to read, my brain hurts!):

> UPDATE, DELETE, SELECT FOR UPDATE, and SELECT FOR SHARE commands behave the same as SELECT in terms of searching for target rows: they will only find target rows that were committed as of the command start time. However, such a target row might have already been updated (or deleted or locked) by another concurrent transaction by the time it is found. In this case, the would-be updater will wait for the first updating transaction to commit or roll back (if it is still in progress).

> If the first updater rolls back, then its effects are negated and the second updater can proceed with updating the originally found row. If the first updater commits .... it will attempt to apply its operation to the updated version of the row.

> The search condition of the command (the WHERE clause) is re-evaluated to see if the updated version of the row still matches the search condition. If so, the second updater proceeds with its operation using the updated version of the row.

If my understanding is correct `READ COMMITTED` will effectively correct the issue by the replay.

Reply:
That said! Synapse's default is actually one better, `REPEATABLE READ`, in which case things are much simpler:

> UPDATE, DELETE, MERGE, SELECT FOR UPDATE, and SELECT FOR SHARE commands behave the same as SELECT in terms of searching for target rows: they will only find target rows that were committed as of the transaction start time. However, such a target row might have already been updated (or deleted or locked) by another concurrent transaction by the time it is found. In this case, the repeatable read transaction will wait for the first updating transaction to commit ... if the first updater commits (and actually updated or deleted the row, not just locked it) then the repeatable read transaction will be rolled back with the message ERROR:  could not serialize access due to concurrent update

Which synapse automatically retries, which would replay the transaction as expected.

Reply:
Thanks! I believe that you're correct (but my brain also hurts)!

Reply:
I'm not sure I agree here. As things stand I don't see why we'd necessarily replay the transactions, as we may not have updated/deleted/locked any of the rows we `SELECT` against (and inserting a new row that would have been picked up by a `SELECT` isn't picked up by postgres except in `SERIALIZABLE` isolation AIUI). 

I think what you want here is to run the selects with a `FOR SHARE` so that they do conflict with each other?

Reply:
I guess most of the time the `UPDATE` will conflict, but if we have two requests to add a rule to the top of the push rules those transactions should conflict but won't?

Reply:
Yeah that makes sense, will add `FOR SHARE` in 👍

Reply:
Wait, we probably need a `FOR UPDATE` instead as we need the `SELECT` statements to conflict with each other and `FOR SHARE` won't do that https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS